### ISSUE TITLE
Question about reproducing mobilenetv1-imagenet-channel pruning-uniform 

### ISSUE THREAD

**Issue by GoldenSpark** [Tencent/PocketFlow#128]
Using default parameters won't get a good performance.The best top1-acc is 59%. Is there any problem with my parameters?
The log is as below:

using Horovod for multi-GPU training
using Horovod for multi-GPU training
-------------------______________________save path:  /data/tf/tf_save_path_2/model.ckpt
-------------------______________________save path:  /data/tf/tf_save_path_2/model.ckpt
using Horovod for multi-GPU training
-------------------______________________save path:  /data/tf/tf_save_path_2/model.ckpt
/data/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
INFO:tensorflow:FLAGS:
INFO:tensorflow:data_disk: local
INFO:tensorflow:data_hdfs_host: None
INFO:tensorflow:data_dir_local: /data/TFRecord/
INFO:tensorflow:data_dir_hdfs: None
INFO:tensorflow:cycle_length: 4
INFO:tensorflow:nb_threads: 8
INFO:tensorflow:buffer_size: 1024
INFO:tensorflow:prefetch_size: 8
INFO:tensorflow:nb_classes: 1001
INFO:tensorflow:nb_smpls_train: 1281167
INFO:tensorflow:nb_smpls_val: 10000
INFO:tensorflow:nb_smpls_eval: 50000
INFO:tensorflow:batch_size: 64
INFO:tensorflow:batch_size_eval: 64
INFO:tensorflow:mobilenet_version: 1
INFO:tensorflow:mobilenet_depth_mult: 1.0
INFO:tensorflow:nb_epochs_rat: 1.0
INFO:tensorflow:lrn_rate_init: 0.045
INFO:tensorflow:batch_size_norm: 96.0
INFO:tensorflow:momentum: 0.9
INFO:tensorflow:loss_w_dcy: 4e-05
INFO:tensorflow:model_http_url: https://api.ai.tencent.com/pocketflow
INFO:tensorflow:summ_step: 100
INFO:tensorflow:save_step: 10000
INFO:tensorflow:save_path: /data/tf/tf_save_path_2/model.ckpt
INFO:tensorflow:save_path_eval: /data/tf/tf_save_path_dst_eval/model.ckpt
INFO:tensorflow:enbl_dst: True
INFO:tensorflow:enbl_warm_start: True
INFO:tensorflow:loss_w_dst: 4.0
INFO:tensorflow:tempr_dst: 4.0
INFO:tensorflow:save_path_dst: /data/tf/tf_save_path_dst/model.ckpt
INFO:tensorflow:ddpg_actor_depth: 2
INFO:tensorflow:ddpg_actor_width: 64
INFO:tensorflow:ddpg_critic_depth: 2
INFO:tensorflow:ddpg_critic_width: 64
INFO:tensorflow:ddpg_noise_type: param
INFO:tensorflow:ddpg_noise_prtl: tdecy
INFO:tensorflow:ddpg_noise_std_init: 1.0
INFO:tensorflow:ddpg_noise_dst_finl: 0.01
INFO:tensorflow:ddpg_noise_adpt_rat: 1.03
INFO:tensorflow:ddpg_noise_std_finl: 1e-05
INFO:tensorflow:ddpg_rms_eps: 0.0001
INFO:tensorflow:ddpg_tau: 0.01
INFO:tensorflow:ddpg_gamma: 0.9
INFO:tensorflow:ddpg_lrn_rate: 0.001
INFO:tensorflow:ddpg_loss_w_dcy: 0.0
INFO:tensorflow:ddpg_record_step: 1
INFO:tensorflow:ddpg_batch_size: 64
INFO:tensorflow:ddpg_enbl_bsln_func: True
INFO:tensorflow:ddpg_bsln_decy_rate: 0.95
INFO:tensorflow:ws_save_path: ./models_ws/model.ckpt
INFO:tensorflow:ws_prune_ratio: 0.75
INFO:tensorflow:ws_prune_ratio_prtl: optimal
INFO:tensorflow:ws_nb_rlouts: 200
INFO:tensorflow:ws_nb_rlouts_min: 50
INFO:tensorflow:ws_reward_type: single-obj
INFO:tensorflow:ws_lrn_rate_rg: 0.03
INFO:tensorflow:ws_nb_iters_rg: 20
INFO:tensorflow:ws_lrn_rate_ft: 0.0003
INFO:tensorflow:ws_nb_iters_ft: 400
INFO:tensorflow:ws_nb_iters_feval: 25
INFO:tensorflow:ws_prune_ratio_exp: 3.0
INFO:tensorflow:ws_iter_ratio_beg: 0.1
INFO:tensorflow:ws_iter_ratio_end: 0.5
INFO:tensorflow:ws_mask_update_step: 500.0
INFO:tensorflow:cp_lasso: True
INFO:tensorflow:cp_quadruple: False
INFO:tensorflow:cp_reward_policy: accuracy
INFO:tensorflow:cp_nb_points_per_layer: 10
INFO:tensorflow:cp_nb_batches: 30
INFO:tensorflow:cp_prune_option: uniform
INFO:tensorflow:cp_prune_list_file: ratio.list
INFO:tensorflow:cp_channel_pruned_path: /data/tf/tf_cp_channel_pruned_path/pruned_model.ckpt
INFO:tensorflow:cp_best_path: /data/tf/tf_cp_best_path/best_model.ckpt
INFO:tensorflow:cp_original_path: /data/tf/tf_cp_original_path/original_model.ckpt
INFO:tensorflow:cp_preserve_ratio: 0.5
INFO:tensorflow:cp_uniform_preserve_ratio: 0.5
INFO:tensorflow:cp_noise_tolerance: 0.15
INFO:tensorflow:cp_lrn_rate_ft: 0.0001
INFO:tensorflow:cp_nb_iters_ft_ratio: 0.2
INFO:tensorflow:cp_finetune: False
INFO:tensorflow:cp_retrain: False
INFO:tensorflow:cp_list_group: 1000
INFO:tensorflow:cp_nb_rlouts: 200
INFO:tensorflow:cp_nb_rlouts_min: 50
INFO:tensorflow:cpg_save_path: /data/tf/tf_cpg_save_path/model.ckpt
INFO:tensorflow:cpg_save_path_eval: /data/tf/tf_cpg_save_path_eval/model.ckpt
INFO:tensorflow:cpg_prune_ratio_type: uniform
INFO:tensorflow:cpg_prune_ratio: 0.5
INFO:tensorflow:cpg_skip_ht_layers: True
INFO:tensorflow:cpg_prune_ratio_file: None
INFO:tensorflow:cpg_lrn_rate_pgd_init: 1e-10
INFO:tensorflow:cpg_lrn_rate_pgd_incr: 1.4
INFO:tensorflow:cpg_lrn_rate_pgd_decr: 0.7
INFO:tensorflow:cpg_lrn_rate_adam: 0.01
INFO:tensorflow:cpg_nb_iters_layer: 1000
INFO:tensorflow:dcp_save_path: /data/tf/tf_dcp_save_path/model.ckpt
INFO:tensorflow:dcp_save_path_eval: /data/tf/tf_dcp_save_path_eval/model.ckpt
INFO:tensorflow:dcp_prune_ratio: 0.5
INFO:tensorflow:dcp_nb_stages: 3
INFO:tensorflow:dcp_lrn_rate_adam: 0.001
INFO:tensorflow:dcp_nb_iters_block: 10000
INFO:tensorflow:dcp_nb_iters_layer: 500
INFO:tensorflow:uql_equivalent_bits: 4
INFO:tensorflow:uql_nb_rlouts: 200
using Horovod for multi-GPU training
INFO:tensorflow:uql_w_bit_min: 2
INFO:tensorflow:uql_w_bit_max: 8
INFO:tensorflow:uql_tune_layerwise_steps: 100
INFO:tensorflow:uql_tune_global_steps: 2000
INFO:tensorflow:uql_tune_save_path: ./rl_tune_models/model.ckpt
INFO:tensorflow:uql_tune_disp_steps: 300
INFO:tensorflow:uql_enbl_random_layers: True
INFO:tensorflow:uql_enbl_rl_agent: False
INFO:tensorflow:uql_enbl_rl_global_tune: True
INFO:tensorflow:uql_enbl_rl_layerwise_tune: False
INFO:tensorflow:uql_weight_bits: 4
INFO:tensorflow:uql_activation_bits: 32
INFO:tensorflow:uql_use_buckets: False
INFO:tensorflow:uql_bucket_size: 256

**Comment by jiaxiang-wu** [comment]
Hi, can you post the execution command?

**Comment by GoldenSpark** [comment]
Hey @jiaxiang-wu ，given a lower learning_rate_init and a higher cp_uniform_preserve_ratio, the top1-acc reaches to 62.6% now. What I actually want to know is how to reach a 66.5% top1-acc with 50% uniform_pruning ratio. Could you tell me the hyper-parameters? Thanks a lot.

**Comment by psyyz10** [comment]
Hi @GoldenSpark , if you wants the FLOPs of pruned model is 50% of the original model by uniform pruning, you should set ` --uniform_preserve_ratio=0.69` and also `--enbl_dst=True`.

**Comment by jiaxiang-wu** [comment]
@GoldenSpark 
We need to clarify that the "pruning ratio" previously used in our table refers to the overall FLOPs, instead of each layer's parameters. Actually, to reduce the FLOPs to 50% of the original model, each layer only needs to prune ~31% channels (as mentioned by @psyyz10 ). We have clarified this in PR #137. Sorry for the confusion.

**Comment by GoldenSpark** [comment]
Hey @psyyz10 @jiaxiang-wu, thaks for your reply, I have noticed the relations between pruning-ratio and FLOPs. But using the default setting except changing pruning-ratio to 0.707 can not reproducing 65.5% top1-acc. Should I optimize the training hyper-parameters?

**Comment by jiaxiang-wu** [comment]
Can you post the execution command for producing 62.8% top-1 accuracy in your experiment?

**Comment by GoldenSpark** [comment]
@jiaxiang-wu  

#!/bin/bash
# default arguments
nb_gpus=8

py_script="nets/mobilenet_at_ilsvrc12_run.py"

LOG_FILE=/data/log/mobilenetv1_img_channel_uni_0.707_retrain_0.106_dst.log

LOG_FILE=/data/log/mobilenetv1_img_cp_uni_0.707_retrain_0.106_dst.log

extra_args="--learner channel \
            --batch_size_eval 64 \
            --cp_uniform_preserve_ratio 0.707107 \
            --cp_prune_option uniform \
            --cp_finetune=False \
            --cp_retrain=True \
            --lrn_rate_init 0.02 \
            --enbl_dst=True \
            --enbl_warm_start=True \
            --data_dir_local /data/TFRecord/ \
            --model_http_url https://api.ai.tencent.com/pocketflow" 

# obtain list of idle GPUs
idle_gpus=`python utils/get_idle_gpus.py ${nb_gpus}`
export CUDA_VISIBLE_DEVICES=${idle_gpus}

# re-create the logging directory
rm -rf logs && mkdir logs

# execute the specified Python script with one or more GPUs
cp -v ${py_script} main.py
if [ ${nb_gpus} -eq 1 ]; then
  echo "multi-GPU training disabled"
  python main.py ${extra_args}
elif [ ${nb_gpus} -le 8 ]; then
  echo "multi-GPU training enabled"
 options="-np ${nb_gpus} -H localhost:${nb_gpus} -bind-to none -map-by slot
     -x NCCL_DEBUG=INFO -x NCCL_SOCKET_IFNAME=eth0 -x NCCL_IB_DISABLE=1
     -x LD_LIBRARY_PATH --mca btl_tcp_if_include eth0"

  echo ${nb_gpus}
  echo ${LOG_FILE}
  echo "extra_args: ${extra_args}"
  mpirun ${options} python main.py --enbl_multi_gpu ${extra_args} 2>&1 | tee $LOG_FILE
fi

**Comment by jiaxiang-wu** [comment]
We use the following execution command:
``` bash
$ ./scripts/run_seven.sh nets/mobilenet_at_ilsvrc12_run.py \
    -n=8 -j=pocketflow-mbn-v1-cp-0-69 \
    --learner channel \
    --enbl_dst \
    --cp_prune_option uniform \
    --cp_uniform_preserve_ratio 0.69 \
    --batch_size_eval 64
```
and the final top-1 accuracy is 67.1% and top-5 accuracy is 87.4%.

The full log is attached here:
[mobilenet-cp.log](https://github.com/Tencent/PocketFlow/files/2678469/mobilenet-cp.log)

**Comment by GoldenSpark** [comment]
Hey @jiaxiang-wu ， Thanks for suppling the log. I have found in my log that top1-acc following every training of 100 iters is abnormal but finally lead to a higher val-top1-acc  which is seems to be nomal. 
### In fact, compared with yours, the mean value of train-top1-acc is so low and the final val-top1-acc is also lower than yours.

Here is my log：
[mobilenetv1_img_channel_uni_0.707_retrain_0.24_dst.log](https://github.com/Tencent/PocketFlow/files/2679280/mobilenetv1_img_channel_uni_0.707_retrain_0.24_dst.log)
[resnet_imagenet_w6a32_no_RL_uniform_quantization.log](https://github.com/Tencent/PocketFlow/files/2679296/resnet_imagenet_w6a32_no_RL_uniform_quantization.log)

**Comment by jiaxiang-wu** [comment]
Can you test with the execution command I provided earlier, on the latest version of PocketFlow code?

**Comment by GoldenSpark** [comment]
@jiaxiang-wu  
There is something wrong with my TFRecords data. I change the shards number of thraning data from 4096 to 1024 and reconvert  the ImageNet to TFRecord.  And now the train-top1-acc is nomal.  
Thanks for your help!
